File: src/main/java/net/bati/guilib/utils/DrawHelper.java
package net.bati.guilib.utils;

import com.mojang.blaze3d.systems.RenderSystem;
import net.minecraft.client.render.*;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.Matrix4f;
import net.minecraft.util.math.Vec3d;
import org.lwjgl.opengl.GL11;


public class DrawHelper {
    public static void drawRectangle(Identifier texture, double x, double y, int u, int v, double width, double height, double scale, int imageWidth, int imageHeight) {
        drawRectangle(texture,x, y, u, v, width, height,scale,imageWidth, imageHeight, null);
    }


    public static void drawRectangle(Identifier texture,double x, double y, int u, int v, double width, double height,double scale, int imageWidth, int imageHeight, Matrix4f matrix)
    {
        RenderSystem.setShader(GameRenderer::getPositionTexShader);
        RenderSystem.setShaderTexture(0, texture);
        BufferBuilder bufferBuilder = Tessellator.getInstance().getBuffer();
        double minU = (double)u / (double)imageWidth;
        double maxU = (double)(u + width) / (double)imageWidth;
        double minV = (double)v / (double)imageHeight;
        double maxV = (double)(v + height) / (double)imageHeight;
        RenderSystem.enableBlend();
        RenderSystem.defaultBlendFunc();
        RenderSystem.enableDepthTest();
        bufferBuilder.begin(VertexFormat.DrawMode.QUADS, VertexFormats.POSITION_TEXTURE);
        bufferBuilder.vertex(matrix,(float)(x + scale*(double)width),(float)( y + scale*(double)height), 0).texture((float)maxU,(float)maxV).next();
        bufferBuilder.vertex(matrix, (float)(x + scale*(double)width), (float)y, 0).texture((float)maxU,(float)minV).next();
        bufferBuilder.vertex(matrix, (float)x, (float)y, 0).texture((float)minU,(float)minV).next();
        bufferBuilder.vertex(matrix, (float)x, (float)(y + scale*(double)height),0).texture((float)minU,(float)maxV).next();
        bufferBuilder.end();
        BufferRenderer.draw(bufferBuilder);

        RenderSystem.disableBlend();
    }
    public static void drawRectangle(Identifier texture,double x, double y, int u, int v, double width, double height,double scalex, double scaley, int imageWidth, int imageHeight, Matrix4f matrix)
    {

        RenderSystem.setShader(GameRenderer::getPositionTexShader);
        RenderSystem.setShaderTexture(0, texture);
        BufferBuilder bufferBuilder = Tessellator.getInstance().getBuffer();
        double minU = (double)u / (double)imageWidth;
        double maxU = (double)(u + width) / (double)imageWidth;
        double minV = (double)v / (double)imageHeight;
        double maxV = (double)(v + height) / (double)imageHeight;
        RenderSystem.enableBlend();
        RenderSystem.defaultBlendFunc();
        RenderSystem.enableDepthTest();
        bufferBuilder.begin(VertexFormat.DrawMode.QUADS, VertexFormats.POSITION_TEXTURE);
        bufferBuilder.vertex(matrix,(float)(x + scalex*(double)width),(float)( y + scaley*(double)height), 0).texture((float)maxU,(float)maxV).next();
        bufferBuilder.vertex(matrix, (float)(x + scalex*(double)width), (float)y, 0).texture((float)maxU,(float)minV).next();
        bufferBuilder.vertex(matrix, (float)x, (float)y, 0).texture((float)minU,(float)minV).next();
        bufferBuilder.vertex(matrix, (float)x, (float)(y + scaley*(double)height),0).texture((float)minU,(float)maxV).next();
        bufferBuilder.end();
        BufferRenderer.draw(bufferBuilder);

        RenderSystem.disableBlend();
    }

    public static void fillGradient(MatrixStack matrices, float startX, float startY, float endX, float endY, int colorStart, float a, int z) {
        RenderSystem.disableTexture();
        RenderSystem.enableBlend();
        RenderSystem.defaultBlendFunc();
        RenderSystem.setShader(GameRenderer::getPositionColorShader);
        Tessellator tessellator = Tessellator.getInstance();
        BufferBuilder bufferBuilder = tessellator.getBuffer();
        bufferBuilder.begin(VertexFormat.DrawMode.QUADS, VertexFormats.POSITION_COLOR);
        fillGradient(matrices.peek().getPositionMatrix(), bufferBuilder, startX, startY, endX, endY, z, colorStart, a);
        tessellator.draw();
        RenderSystem.disableBlend();
        RenderSystem.enableTexture();
    }

    public static void fillGradient(Matrix4f matrix, BufferBuilder builder, float startX, float startY, float endX, float endY, int z, int colorStart, float a) {
        Vec3d color = Vec3d.unpackRgb(colorStart);
        builder.vertex(matrix, (float)endX, (float)startY, (float)z).color((float)color.x, (float)color.y, (float)color.z, a).next();
        builder.vertex(matrix, (float)startX, (float)startY, (float)z).color((float)color.x, (float)color.y, (float)color.z, a).next();
        builder.vertex(matrix, (float)startX, (float)endY, (float)z).color((float)color.x, (float)color.y, (float)color.z, a).next();
        builder.vertex(matrix, (float)endX, (float)endY, (float)z).color((float)color.x, (float)color.y, (float)color.z, a).next();

    }



    public static void hexColor(int pColor, float alpha) {
        float h2 = (pColor >> 16 & 0xFF) / 255.0F;
        float h3 = (pColor >> 8 & 0xFF) / 255.0F;
        float h4 = (pColor & 0xFF) / 255.0F;
        float h1 = 1.0F;
        RenderSystem.setShaderColor(h1 * h2, h1 * h3, h1 * h4, alpha);
    }

    public static void drawFilledArc(int xCenter, int yCenter, int radius, double startDegrees, double finishDegrees, int color, float opacity, MatrixStack matrices) {
        matrices.push();
        RenderSystem.disableTexture();
        RenderSystem.enableBlend();
        hexColor(color, opacity);
        GL11.glBegin(6);
        GL11.glVertex2d(xCenter, yCenter);
        double i;
        for (i = startDegrees; i <= finishDegrees; i += 0.05D) {
            double theta = Math.PI*2 * i / 360.0D;
            double dotX = xCenter + Math.sin(theta) * radius;
            double dotY = yCenter + Math.cos(theta) * radius;
            GL11.glVertex2d(dotX , dotY);
        }
        GL11.glEnd();
        RenderSystem.enableTexture();
        RenderSystem.disableBlend();
        matrices.pop();
    }
}

File: src/main/kotlin/net/bati/guilib/utils/DrawHelper.kt
package net.bati.guilib.utils

import com.mojang.blaze3d.systems.RenderSystem
import net.bati.guilib.gui.components.Widget
import net.minecraft.client.render.*
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.util.Identifier
import net.minecraft.util.math.Matrix4f
import net.minecraft.util.math.Vec3d
import org.jetbrains.annotations.NotNull
import org.jetbrains.annotations.Nullable

object DrawHelper {
    @JvmOverloads
    @JvmStatic
    fun drawRectangle(
        texture: Identifier?,
        x: Double,
        y: Double,
        u: Int,
        v: Int,
        width: Double,
        height: Double,
        scale: Double = 1.0,
        imageWidth: Int = 256,
        imageHeight: Int = 256,
        matrix: Matrix4f? = null,
        z: Float = 0.0F
    ) {
        RenderSystem.setShader { GameRenderer.getPositionTexShader() }
        RenderSystem.setShaderTexture(0, texture)
        val bufferBuilder = Tessellator.getInstance().buffer
        val minU = u.toDouble() / imageWidth.toDouble()
        val maxU = (u + width) / imageWidth.toDouble()
        val minV = v.toDouble() / imageHeight.toDouble()
        val maxV = (v + height) / imageHeight.toDouble()
        RenderSystem.enableBlend()
        RenderSystem.defaultBlendFunc()
        RenderSystem.enableDepthTest()
        bufferBuilder.begin(VertexFormat.DrawMode.QUADS, VertexFormats.POSITION_TEXTURE)
        bufferBuilder.vertex(matrix, (x + scale * width).toFloat(), (y + scale * height).toFloat(), z)
            .texture(maxU.toFloat(), maxV.toFloat()).next()
        bufferBuilder.vertex(matrix, (x + scale * width).toFloat(), y.toFloat(), z)
            .texture(maxU.toFloat(), minV.toFloat()).next()
        bufferBuilder.vertex(matrix, x.toFloat(), y.toFloat(), z).texture(minU.toFloat(), minV.toFloat()).next()
        bufferBuilder.vertex(matrix, x.toFloat(), (y + scale * height).toFloat(), z)
            .texture(minU.toFloat(), maxV.toFloat()).next()
        bufferBuilder.end()
        BufferRenderer.draw(bufferBuilder)
        RenderSystem.disableBlend()
    }

    @JvmStatic
    fun drawRectangle2(
        texture: Identifier?,
        x: Double,
        y: Double,
        u: Int,
        v: Int,
        width: Double,
        height: Double,
        scalex: Double,
        scaley: Double,
        imageWidth: Int,
        imageHeight: Int,
        matrix: Matrix4f?
    ) {
        RenderSystem.setShader { GameRenderer.getPositionTexShader() }
        RenderSystem.setShaderTexture(0, texture)
        val bufferBuilder = Tessellator.getInstance().buffer
        val minU = u.toDouble() / imageWidth.toDouble()
        val maxU = (u + width) / imageWidth.toDouble()
        val minV = v.toDouble() / imageHeight.toDouble()
        val maxV = (v + height) / imageHeight.toDouble()
        RenderSystem.enableBlend()
        RenderSystem.defaultBlendFunc()
        RenderSystem.enableDepthTest()
        bufferBuilder.begin(VertexFormat.DrawMode.QUADS, VertexFormats.POSITION_TEXTURE)
        bufferBuilder.vertex(matrix, (x + scalex * width).toFloat(), (y + scaley * height).toFloat(), 0f)
            .texture(maxU.toFloat(), maxV.toFloat()).next()
        bufferBuilder.vertex(matrix, (x + scalex * width).toFloat(), y.toFloat(), 0f)
            .texture(maxU.toFloat(), minV.toFloat()).next()
        bufferBuilder.vertex(matrix, x.toFloat(), y.toFloat(), 0f).texture(minU.toFloat(), minV.toFloat()).next()
        bufferBuilder.vertex(matrix, x.toFloat(), (y + scaley * height).toFloat(), 0f)
            .texture(minU.toFloat(), maxV.toFloat()).next()
        bufferBuilder.end()
        BufferRenderer.draw(bufferBuilder)
        RenderSystem.disableBlend()
    }

    @JvmStatic
    fun fillGradient(
        matrices: MatrixStack,
        startX: Float,
        startY: Float,
        endX: Float,
        endY: Float,
        colorStart: Int,
        a: Float,
        z: Int
    ) {
        RenderSystem.disableTexture()
        RenderSystem.enableBlend()
        RenderSystem.defaultBlendFunc()
        RenderSystem.setShader { GameRenderer.getPositionColorShader() }
        val tessellator = Tessellator.getInstance()
        val bufferBuilder = tessellator.buffer
        bufferBuilder.begin(VertexFormat.DrawMode.QUADS, VertexFormats.POSITION_COLOR)
        fillGradient(matrices.peek().positionMatrix, bufferBuilder, startX, startY, endX, endY, z, colorStart, a)
        tessellator.draw()
        RenderSystem.disableBlend()
        RenderSystem.enableTexture()
    }

    @JvmStatic
    fun fillGradient(
        matrix: Matrix4f?,
        builder: BufferBuilder,
        startX: Float,
        startY: Float,
        endX: Float,
        endY: Float,
        z: Int,
        colorStart: Int,
        a: Float
    ) {
        val color = Vec3d.unpackRgb(colorStart)
        builder.vertex(matrix, endX, startY, z.toFloat())
            .color(color.x.toFloat(), color.y.toFloat(), color.z.toFloat(), a).next()
        builder.vertex(matrix, startX, startY, z.toFloat())
            .color(color.x.toFloat(), color.y.toFloat(), color.z.toFloat(), a).next()
        builder.vertex(matrix, startX, endY, z.toFloat())
            .color(color.x.toFloat(), color.y.toFloat(), color.z.toFloat(), a).next()
        builder.vertex(matrix, endX, endY, z.toFloat())
            .color(color.x.toFloat(), color.y.toFloat(), color.z.toFloat(), a).next()
    }
    @JvmStatic
    fun drawWithPivot(@NotNull matrices : MatrixStack, x : Int, y : Int, width : Float, height : Float, size : Float, delta : Float, @NotNull pivot: Pivot, @NotNull drawCallback : Callback.DrawableBasic) {
        drawWithPivot(matrices, x, y, width, height, size, delta, pivot, drawCallback, null)
    }
    @JvmStatic
    fun drawWithPivot(@NotNull matrices : MatrixStack, x : Int, y : Int, width : Float, height : Float, size : Float, delta : Float, @NotNull pivot: Pivot, @NotNull drawCallback : Callback.DrawableBasic, @Nullable transformationCallback : Callback.DrawableBasic?) {
        val offsetX: Float = pivot.getX(width)
        val offsetY: Float = pivot.getY(height)

        matrices.push()

        matrices.translate(-offsetX.toDouble(), -offsetY.toDouble(), 0.0)
        matrices.push()
        matrices.translate((x + offsetX).toDouble(), (y + offsetY).toDouble(), 0.0)
        matrices.scale(size, size, size)
        transformationCallback?.draw()
        matrices.translate(-offsetX.toDouble(), -offsetY.toDouble(), 0.0)

        drawCallback.draw()

        matrices.pop()
        matrices.pop()

    }
}